helix-core/src/main/java/org/apache/helix/manager/zk/CallbackEventExecutor.java

```diff
@@ -0,0 +1,122 @@
+ * 2. The previous call back event handling process is finished in thread pool.
+ */
+
+public class CallbackEventExecutor {
+  private static Logger logger = LoggerFactory.getLogger(CallbackHandler.class);
+
+  private DedupEventBlockingQueue<NotificationContext.Type, NotificationContext>
+      _callBackEventQueue;
+  private final HelixManager _manager;
+  private Future _futureCallBackProcessEvent = null;
+  private final ThreadPoolExecutor _threadPoolExecutor;
+
+  public CallbackEventExecutor(HelixManager manager) {
+    _callBackEventQueue = new DedupEventBlockingQueue<>();
+    _manager = manager;
+    _threadPoolExecutor = CallbackEventTPFactory.getOrCreateThreadPool(manager.hashCode());
+  }
+
+  class CallbackProcessor implements Runnable {
+    private CallbackHandler _handler;
+    protected String _processorName;
+    NotificationContext _event;
+
+    public CallbackProcessor(CallbackHandler handler, NotificationContext event) {
+      _processorName = _manager.getClusterName() + "CallbackProcessor@" + Integer
+          .toHexString(handler.hashCode());
+      _handler = handler;
+      _event = event;
+    }
+
+    @Override
+    public void run() {
+      try {
+        _handler.invoke(_event);
+      } catch (ZkInterruptedException e) {
+        logger.warn(_processorName + " thread caught a ZK connection interrupt", e);
+      } catch (ThreadDeath death) {
+        throw death;
+      } catch (Throwable t) {
+        logger.error(_processorName + " thread failed while running " + _processorName, t);
+      }
+      submitPendingHandleCallBackEventToManagerThreadPool(_handler);
+    }
+  }
+
+  public void submitEventToEvecutor(NotificationContext.Type eventType, NotificationContext event,
+      CallbackHandler handler) {
+    synchronized (_callBackEventQueue) {
+      if (_futureCallBackProcessEvent == null || _futureCallBackProcessEvent.isDone()) {
+        _futureCallBackProcessEvent =
+            _threadPoolExecutor.submit(new CallbackProcessor(handler, event));
+      } else {
+        _callBackEventQueue.put(eventType, event);
+      }
+    }
+  }
+
+  private void submitPendingHandleCallBackEventToManagerThreadPool(CallbackHandler handler) {
+    synchronized (_callBackEventQueue) {
+      if (_callBackEventQueue.size() != 0) {
+        try {
+          NotificationContext event = _callBackEventQueue.take();
+          _futureCallBackProcessEvent =
+              _threadPoolExecutor.submit(new CallbackProcessor(handler, event));
+        } catch (InterruptedException e) {
+          e.printStackTrace();
+        }
+      }
+    }
+  }
+
+  public void reset() {
+    synchronized (_callBackEventQueue) {
+      _callBackEventQueue.clear();
+      if (_futureCallBackProcessEvent != null) {
+        _futureCallBackProcessEvent.cancel(false);
+      }
+    }
+  }
+
+  public void unregisterFromFactory() {
+    reset();
+    CallbackEventTPFactory.unregisterEventProcessor(_manager.hashCode());
```

Comment:
You will need to set _threadPoolExecutor = null too.

Moreover, to ensure it is safe after unregister, the executor shall reject any new task submission after this call.